### DISCUSSION TITLE
Complete the square

### DISCUSSION THREAD

**Discussion by omasoud** [sympy/sympy#24449] [Q&A] [answered]
I cannot seem to be able to simplify some (not too complicated) expressions. I'm using sympy 1.11.1.

```python
from sympy import expand, factor, powsimp, simplify, sqrt, symbols, collect, together
x= symbols('x')
s1 = (sqrt(x**2-x+1)+1)**2
s1
```
This gives:
![image](https://user-images.githubusercontent.com/16273458/210125113-56dd0ac1-76d6-4267-a610-5ce760ef1327.png)

```python
s2 = expand(s1)
s2
```
This gives:
![image](https://user-images.githubusercontent.com/16273458/210125125-9f969b0f-f465-48ee-bd98-28de97382c44.png)

I cannot get the simplified expression (`s1`) back from `s2` no matter what I do. I tried `factor()`, `simplify()`, `powsimp()`, `collect()`, `together()`.

This seems simple enough for there to be a way to go from s2 to s1. Any ideas?
Thanks!

**Comment by oscarbenjamin** [comment, marked as answer]
There isn't really a good way to do this with symbolic radicals but you can use explicit substitutions to eliminate the radical and then factor:
```python
In [15]: s2.subs(x**2, y+x-1)
Out[15]: 2⋅√y + y + 1

In [16]: s2.subs(x**2, y+x-1).subs(sqrt(y), z)
Out[16]: 
 2          
z  + 2⋅z + 1

In [17]: s2.subs(x**2, y+x-1).subs(sqrt(y), z).factor()
Out[17]: 
       2
(z + 1) 

In [18]: s2.subs(x**2, y+x-1).subs(sqrt(y), z).factor().subs(z, sqrt(x**2-x+1))
Out[18]: 
                     2
⎛   ____________    ⎞ 
⎜  ╱  2             ⎟ 
⎝╲╱  x  - x + 1  + 1⎠ 
```
Ideally something like `factor(extension=True)` should work here.

**Comment by omasoud** [comment]
Thanks. That's a useful workaround. Hopefully, simplification functionality can be improved for radicals in the future.
Here's a minimal example with the same issue:
```python
from sympy import expand, factor, simplify, sqrt, symbols
x= symbols('x')
y1 = (sqrt(x)+1)**2
print(y1)
print(y1.expand())
print(y1.expand().simplify())
print(y1.expand().factor())
print(y1.expand().factor(extension=True))
```

Prints:
```
(sqrt(x) + 1)**2
2*sqrt(x) + x + 1
2*sqrt(x) + x + 1
2*sqrt(x) + x + 1
2*sqrt(x) + x + 1
```
